packages/agent/src/canisterStatus/index.ts

```diff
@@ -209,6 +234,69 @@ export const request = async (options: {
   return status;
 };
 
+export const fetchNodeKeys = (
+  certificate: ArrayBuffer,
+  canisterId: Principal,
+  root_key?: ArrayBuffer | Uint8Array,
+): SubnetStatus => {
+  if (!canisterId._isPrincipal) {
+    throw new Error('Invalid canisterId');
+  }
+  const cert = Cbor.decode(new Uint8Array(certificate)) as Cert;
+  const tree = cert.tree;
+  let delegation = cert.delegation;
+  let subnetId: Principal;
+  if (delegation && delegation.subnet_id) {
+    subnetId = Principal.fromUint8Array(new Uint8Array(delegation.subnet_id));
+  }
+
+  // On local replica, with System type subnet, there is no delegation
+  else if (!delegation && typeof root_key !== 'undefined') {
+    subnetId = Principal.selfAuthenticating(new Uint8Array(root_key));
+    delegation = {
+      subnet_id: subnetId.toUint8Array(),
+      certificate: new ArrayBuffer(0),
+    };
+  }
+  // otherwise use default NNS subnet id
+  else {
+    subnetId = Principal.selfAuthenticating(
+      Principal.fromText(
+        'tdb26-jop6k-aogll-7ltgs-eruif-6kk7m-qpktf-gdiqx-mxtrf-vb5e6-eqe',
+      ).toUint8Array(),
+    );
+    delegation = {
+      subnet_id: subnetId.toUint8Array(),
+      certificate: new ArrayBuffer(0),
+    };
+  }
+
+  const canisterInRange = check_canister_ranges({ canisterId, subnetId, tree });
+  if (!canisterInRange) {
+    throw new Error('Canister not in range');
+  }
+
+  const nodeTree = lookup_path(['subnet', delegation?.subnet_id as ArrayBuffer, 'node'], tree);
+  const nodeForks = flatten_forks(nodeTree as HashTree) as HashTree[];
+  nodeForks.length;
+  const nodeKeys = new Map<string, DerEncodedPublicKey>();
+  nodeForks.forEach(fork => {
+    Object.getPrototypeOf(new Uint8Array(fork[1] as ArrayBuffer));
+    const node_id = Principal.from(new Uint8Array(fork[1] as ArrayBuffer)).toText();
```

Comment:
We got error:
`index.js:403 Expected to find result for path subnet, but instead found nothing.
(anonymous) @ index.js:403
index.js:403 Error: Impossible to convert {"0":2,"1":229,"2":223,"3":21,"4":96,"5":194,"6":199,"7":168,"8":58,"9":105,"10":151,"11":120,"12":17,"13":148,"14":220,"15":20,"16":146,"17":229,"18":85,"19":55,"20":122,"21":97,"22":23,"23":146,"24":236,"25":191,"26":229,"27":1,"28":2} to Principal.`
Should here be`Principal.fromUint8Array()`? I am curious why it is not detected by tests.

Reply:
I shipped a change in this PR that updates the `Principal.from` method as well, so that it will cover `Uint8Arrays`. You'll need to pack and install `@dfinity/principal`